Feature/signal issue

TeraTian opened the pull request:
code modification for issue #175
https://github.com/mehdy/keepalived-exporter/issues/175

mehdy reviewed (changes requested):
Generally, looks like a good idea. Please fix the lining issue and then we can merge it.

TeraTian commented:
I have updated the file I modified.
But after I run the `golangci-lint` locally, I find that some lint error comes from other files that I didn't touch. 
I have also tried to use `gofumpt`, `gci` command to fix them, but the commands just add a lot of empty newlines. 
I'm not sure if I should include those modification.
Could you please check it?

mehdy commented:
Sure! Let's see if #178 will fix it.

clwluvw commented on internal/types/container/keepalived_container_collector_host.go:
```diff
@@ -135,12 +140,41 @@ func (k *KeepalivedContainerCollectorHost) sigNum(sigString string) syscall.Sign
+	data, err := os.ReadFile(k.pidPath)
 	if err != nil {
-		logrus.WithError(err).WithField("signal", int(signal)).Error("Failed to send signal")
+		logrus.WithField("path", k.pidPath).WithError(err).Error("Can't find keepalived pid. Use default kill command")
```
I guess this fallback would be annoying to be logged here and somehow under control - maybe better to have it as an option and also not falling back on every error?

TeraTian replied:
This modification requires users to add a new mount at "/var/run/keepalived.pid". 
To ensure compatibility with previous versions, I’m keeping a fallback. 
If someone always uses the 'latest' tag for the Docker image and doesn’t notice this change, the new version will still work with their existing keepalived container.

TeraTian commented:
> Sure! Let's see if #178 will fix it.

Hi, I have merged #178 into my branch. Could you please check if it can be merged into master?

mehdy commented on internal/types/container/keepalived_container_collector_host.go:
```diff
@@ -142,12 +147,41 @@ func (k *KeepalivedContainerCollectorHost) sigNum(sigString string) syscall.Sign
+	// Create the execution instance
+	execIDResp, err := k.dockerCli.ContainerExecCreate(context.Background(), k.containerName, execConfig)
+	if err != nil {
+		logrus.WithError(err).Error("Error creating exec instance")
```
Shouldn't you `return err` here?

TeraTian replied:
Yes! I will fixt it.

mehdy commented on internal/types/container/keepalived_container_collector_host.go:
```diff
@@ -142,12 +147,41 @@ func (k *KeepalivedContainerCollectorHost) sigNum(sigString string) syscall.Sign
+	// Start the execution of the created command
+	err = k.dockerCli.ContainerExecStart(context.Background(), execIDResp.ID, types.ExecStartCheck{})
+	if err != nil {
+		logrus.WithError(err).Error("Error starting exec command")
```
Shouldn't you return err here?

TeraTian replied:
Yes! I will fixt it.

mehdy commented:
> > Sure! Let's see if #178 will fix it.
> 
> Hi, I have merged #178 into my branch. Could you please check if it can be merged into master?

The `Lint` job is now working on master. So probably something in the PR is failing the linter.

TeraTian commented:
Hi mehdy
    I have tried using "gci write --skip-generated -s standard -s default /home/devops/golint/keepalived_container_collector_host.go" to fix the problem, but the result looks not good.
    The gci version is 0.13.4. I think it is the latest version.
    I also tried to fix it manually, but I just can't find where the problem lays.
    Could you please give me some instructions about how do you fix lint problem? I can follow your steps. Thank you very much!

mehdy commented:
It's a formatting issue. So if you just run `go fmt ./...` in the root of the project it should fix it for you.

TeraTian commented:
> It's a formatting issue. So if you just run `go fmt ./...` in the root of the project it should fix it for you.

It works! Thanks a lot. I have tested successfully in my own repository. 

Please run the workflow to check it again, thanks.

mehdy commented on internal/types/container/keepalived_container_collector_host.go:
```diff
@@ -142,9 +147,42 @@ func (k *KeepalivedContainerCollectorHost) sigNum(sigString string) syscall.Sign
+	data, err := os.ReadFile(k.pidPath)
 	if err != nil {
-		logrus.WithError(err).WithField("signal", int(signal)).Error("Failed to send signal")
+		logrus.WithField("path", k.pidPath).WithError(err).Error("Can't find keepalived pid. Send signal to default process")
```
Can you change this log to a `Warn`ing instead?

And please change the "Send signal to default process" to  "Falling back to the default process".

clwluvw replied:
I believe this needs to be a debug log as this will just generate annoying logs for the ones using signals.
also, I'm wondering why we can't go with something like
```go
if k.pidPath != "" {
  // do exec
} else {
  // do signal
}
```

TeraTian replied:
Hi
Because k.pidPath always has the default value "/var/run/keepalived.pid", it can't be judged by `k.pidPath != ""`.
And I think this log is not only a debug log, it can give user some recommendation if the user didn't notice this modification. So I add some more log.
What do you think?

clwluvw replied:
Maybe it's better not to have a default value for that then.

mehdy replied:
@TeraTian @clwluvw What do you think guys? Should we remove the default value for `pidPath`?

clwluvw replied:
I believe the current implementation will confuse people not intending or intending to use it as it issues the other party anyway without respecting the user's input.

TeraTian replied:
I think we should keep this default value. Followings are the questions I ask myself.

**1 Why is the default value set to /var/run/keepalived.pid?**
The exporter isn't meant to run independently; it needs to work alongside keepalived. Since /var/run/keepalived.pid is keepalived’s default PID file location, it makes sense to align with that.

**2 Why have a default value?**
Most users don't want to deal with technical details—they just want the program to work with minimal effort. A default value ensures the program runs correctly without needing manual configuration. If we didn’t have this default, users would have to know the keepalived PID and enter it manually, which adds unnecessary complexity. The simpler, the better.

**3 Does this override user input?**
Not at all. Most users won’t need to change this value unless there’s a specific reason. And if they do, they can easily specify their own value. There’s no restriction on that.

TeraTian replied:
Hi @mehdy, any idea about this?

mehdy replied:
Hi @TeraTian,
Thanks. I believe this is good reasoning. It makes sense to keep the default value for `pidPath`.
I'm ok with this setup but with a couple of tweaks.
1. I actually think this should be an INFO-level log. It shouldn't alert the user with `WARN` or `ERROR` and it also should be visible so I wouldn't go with `DEBUG`.
2. I'd suggest removing this part from the log "It's recommended to explicitly specify the pid.". I'm not sure if we should make this recommendation.

TeraTian replied:
Yes, I agree with you. I have updated the code. Please review it again, thanks!

mehdy reviewed (changes requested):
Let's fix this last point and merge this.
